nexus/src/authn/silos.rs

```diff
@@ -0,0 +1,156 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Silo related authentication types and functions
+
+use crate::db::model::SiloSamlIdentityProvider;
+
+use anyhow::{anyhow, bail, Result};
+use samael::metadata::ContactPerson;
+use samael::metadata::ContactType;
+use samael::metadata::EntityDescriptor;
+use samael::metadata::NameIdFormat;
+use samael::metadata::HTTP_REDIRECT_BINDING;
+use samael::service_provider::ServiceProvider;
+use samael::service_provider::ServiceProviderBuilder;
+
+pub enum SiloIdentityProviderType {
+    Local,
+    Ldap,
+    Saml(Box<SiloSamlIdentityProvider>),
+}
+
+impl SiloSamlIdentityProvider {
+    /// return an error if this SiloSamlIdentityProvider is invalid
+    pub fn validate(&self) -> Result<()> {
```

Comment:
I was interested in trying this as part of the changes in 5f352d05, something like

```diff
diff --git a/nexus/src/external_api/params.rs b/nexus/src/external_api/params.rs
index cc18866a..5834e1dc 100644
--- a/nexus/src/external_api/params.rs
+++ b/nexus/src/external_api/params.rs
@@ -30,7 +30,8 @@ pub struct SiloCreate {
 #[derive(Clone, Debug, Deserialize, Serialize, JsonSchema)]
 pub struct DerEncodedKeyPair {
     /// request signing public certificate (base64 encoded der file)
-    pub public_cert: String,
+    #[serde(deserialize_with = "openssl::x509::X509::from_der")]
+    pub public_cert: openssl::x509::X509,
 
     /// request signing private key (base64 encoded der file)
     pub private_key: String,
```

but that doesn't compile. I'm not sure how to say that the incoming type on the wire is `[u8]`, but attempt to deserialize with that function, and fail if it can't.

Reply:
note I'm aware that it has to be base64 decoded as well, but I thought I could at least get something compiling :)

Reply:
Validate during deserialize done in fa10a10b. Using openssl's types as fields unfortunately does not work because they do not impl JsonSchema.